### ISSUE TITLE
Keras 'Random' preprocessing layers do not work on cloud TPU

### ISSUE THREAD

**Issue by Lipod** [tensorflow/tensorflow#42454]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Simple demo available in colab [here](https://colab.research.google.com/drive/1lZBE7bDvRxRoEijYkx8Zch3JU9xug6V3?usp=sharing).

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Google Colab with TPU

- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
N/A

- TensorFlow installed from (source or binary):
Google Colab

- TensorFlow version (use command below):
2.3.0

- Python version:
3.6.9

- Bazel version (if compiling from source):
N/A
- GCC/Compiler version (if compiling from source):
N/A
- CUDA/cuDNN version:
N/A
- GPU model and memory:
Cloud TPU

** Current Behaviour **
'Random' layers within tf.keras.layers.preprocessing.experimental (e.g. RandomRotation, RandomTranslation) fail with `Detected unsupported operations when trying to compile graph` ... `No registered 'ImageProjectiveTransformV2' OpKernel for XLA_TPU_JIT devices compatible with node` when using a cloud TPU, either in colab or from a VM

** Expected Behaviour **
New layers should use ops that are implemented on TPUs, or at least carry a health warning that they are not (yet) available on TPU.

**Standalone code to reproduce the issue**
See [this colab](https://colab.research.google.com/drive/1lZBE7bDvRxRoEijYkx8Zch3JU9xug6V3?usp=sharing). Or run the following code with a TPU:
```
import tensorflow as tf

# establish connection to TPU
tpu = tf.distribute.cluster_resolver.TPUClusterResolver()
tf.config.experimental_connect_to_cluster(tpu)
tf.tpu.experimental.initialize_tpu_system(tpu)
strategy = tf.distribute.TPUStrategy(tpu)
print('Running on TPU ', tpu.cluster_spec().as_dict()['worker'])  

with strategy.scope():
  # define model layers
  model_input = tf.keras.layers.Input(shape=(224, 224, 1))
  x = tf.keras.layers.experimental.preprocessing.RandomRotation((-0.1, 0.1))(model_input)
  x = tf.keras.layers.MaxPool2D((224, 224))(x)
  x = tf.keras.layers.Flatten()(x)
  x = tf.keras.layers.Dense(1, activation='sigmoid')(x)

  # compile model
  model = tf.keras.Model(inputs=model_input, outputs=x)
  model.summary(line_length=120)
  model.compile(optimizer='adam', loss='binary_crossentropy')

  # generate some random data and fit the model
  images = tf.random.uniform((10, 224, 224, 1))
  labels = tf.zeros((10, 1))
  model.fit(images, labels)

  # predict on the data
  print(model.predict(images))
```

**Other info / logs**
I appreciate that these are 'experimental' features, so I am not sure if this constitutes a bug or a feature request. It would be helpful to indicate if TPU support for these new keras layers is in the pipeline - I have been looking for an image augmentation solution in tf2.0+ for some time, and was hopeful these new layers would be it. Without TPU support the feature has limited use at this time. It would also have been helpful to put a health warning in the docs that TPU support is not yet available, as I would not have wasted my time on this. Someone has recently [asked a question about this](https://stackoverflow.com/questions/63302446/colabtpu-not-supporting-tf-2-3-0-tf-keras-layers-experimental-preprocessing) on StackOverflow, but in the absence of an official reply there I chose to raise it here.

Note that I tried using a VM with tf-nightly (2.4.0-dev20200817) installed, but I get the same error message. Since I can only create TPUs at version 2.3, this is not altogether surprising.

Thanks

**Comment by Saduf2019** [comment]
I am able to replicate this issue, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/7bad14f22c5624458d66b3e6caf0bba8/untitled367.ipynb)
Thanks!

**Comment by Lipod** [comment]
In the meantime, as a workaround I created a standalone 'layer' and incorporated it into my input pipeline instead of in the model, e.g.:
```
augmentor = tf.keras.layers.experimental.preprocessing.RandomRotation((-0.1, 0.1))
ds = ds.map(lambda x, y: (augmentor.call(x), y))
```
This is probably abuse of a keras layer, but it seems to work on a TPU.

**Comment by nikitamaia** [comment]
Hi @Lipod, glad you were able to find a workaround for now. I've changed the label from bug to a docs-feature. Sounds like you're proposing that the docs for keras layers include a warning if they don't work on TPUs. Additionally you're asking for some insight into what keras layers are planned to be supported in the near future. Is that correct?

**Comment by Lipod** [comment]
Hi @nikitamaia ,

I disagree with your decision. Primarily, I would characterise this as a bug. As per my instructions to reproduce, I tried to use a keras feature on a cloud TPU and it failed. Since keras is primary high-level API, and TPUs are one of the key features of tensorflow, I think it is reasonable for me, as an end user, to expect new keras features to work on a TPU. I am asking for someone to fix the bug so that I can use these layers on a TPU.

My reference to updating the documentation was because I was anticipating that you might say that these layers are not supported on a TPU (I probably overthought this). To an end user like me reporting his first bug, you might already know it doesn't work on a TPU and either be planning to introduce it later or never support it. If either of those are true then you could have put that in the documentation and saved users like me the bother of trying to use an unsupported feature. As it is, from the comments of your colleagues above, it looks like this is a straight up bug or oversight. I guess your test coverage for keras does not include running on a TPU.

Regarding my workaround, it is clear that these new layers are intended to be compiled into a keras model and not used in the input pipeline (hence my comment about abusing a keras layer), so it is not ideal.  If you can't make these layers work in a compiled model on a TPU then you might consider making equivalent functions available in the  `tf.image` namespace or similar so that pipeline use is officially supported.

**Comment by nikitamaia** [comment]
Ah okay, thanks for the clarification. I've changed the label back to bug. tf keras preprocessing layers were recently added to the TPU tests and currently RandomRotation and RandomTranslation fail. So this is a known issue, but a bug nonetheless.

Another workaround for the time being is to set `tf.config.set_soft_device_placement(True)` at the beginning of the program. It will run the RandomRotation op on CPU. Obviously not ideal, but it will allow your program to run.

**Comment by nikitamaia** [comment]
Hi @Lipod, providing an update after lots of discussions between the keras and TPU teams. While this originally seemed like a bug, TPU implementation for `ImageProjectiveTransform` (and `V2`) has been determined to be much less efficient than just using CPU. Recommendations here are to either move the preprocessing to your tf.data transformations if you want to use TPUs, or to just use CPU or GPU if you want to use these tf.keras image preprocessing layers. Main takeaway is that supporting `ImageProjectiveTransform` at the TPU level is discouraged due to the implementation being much less efficient than on CPU.

**Comment by Lipod** [comment]
Hi @nikitamaia, thanks for the update. In that case, as an end user, it would help if you could:

1. Expose these augmentations in an API appropriate for use in the `tf.data` pipeline (such as `tf.image`). This would effectively make my workaround 'official', and would separate the ops from all of the scaffold that comes with a keras layer.

2. Update the documentation in the `tf.keras.preprocessing` namespace to warn users that they do not support a TPU, and suggest workarounds.

Thanks

**Comment by nikitamaia** [comment]
Agreed, we do need a way to make this more apparent to users. Will keep this issue open until that information is available somewhere other than just this thread.

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF 2.6.0-dev20210528,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/cafccc4b64d87a241f318502ca1c5e26/untitled58.ipynb#scrollTo=weWesOVfS_R8)..Thanks !

**Comment by sachinprasadhs** [comment]
`tf.keras.layers.experimental.preprocessing.RandomRotation` has been changed to `tf.keras.layers.RandomRotation` and the error message has been improved to show the specific error message. Please find the gist [here](https://colab.sandbox.google.com/gist/sachinprasadhs/b1b8502ae01616a6844e53c07a17a613/random_layers_fail_on_tpu.ipynb) for reference. Thanks!
